controllers/vgmanager.go

```diff
@@ -0,0 +1,127 @@
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type vgManager struct{}
+
+var _ resourceManager = vgManager{}
+
+const (
+	VGManagerUnit = "vg-manager"
+)
+
+func (v vgManager) getName() string {
+	return VGManagerUnit
+}
+
+func (v vgManager) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	unitLogger := r.Log.WithValues("resourceManager", v.getName())
+	// aggregate nodeSelector and tolerations from all deviceClasses
+	nodeSelector, tolerations := extractNodeSelectorAndTolerations(*lvmCluster)
+	// get desired daemonset spec
+	dsTemplate := newVGManagerDaemonset(*lvmCluster)
+	// create desired daemonset or update mutable fields on existing one
+	ds := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dsTemplate.Name,
+			Namespace: dsTemplate.Namespace,
+		},
+	}
+	unitLogger.Info("running CreateOrUpdate")
+	// the anonymous mutate function modifies the daemonset object after fetching it.
+	// if the daemonset does not already exist, it creates it, otherwise, it updates it
+	result, err := ctrl.CreateOrUpdate(ctx, r.Client, ds, func() error {
+		// at creation, deep copy the whole daemonset
+		if ds.CreationTimestamp.IsZero() {
+			dsTemplate.DeepCopyInto(ds)
+			return nil
+		}
+		// if update, update only mutable fields
+
+		// copy selector labels to daemonset and template
+		initMapIfNil(&ds.ObjectMeta.Labels)
+		initMapIfNil(&ds.Spec.Template.ObjectMeta.Labels)
+		for key, value := range dsTemplate.Labels {
+			ds.ObjectMeta.Labels[key] = value
+			ds.Spec.Template.ObjectMeta.Labels[key] = value
+		}
+
+		// containers
+		ds.Spec.Template.Spec.Containers = dsTemplate.Spec.Template.Spec.Containers
+
+		// volumes
+		ds.Spec.Template.Spec.Volumes = dsTemplate.Spec.Template.Spec.Volumes
+
+		// service account
+		// TODO(rohan) get from env
+		ds.Spec.Template.Spec.ServiceAccountName = "lvm-operator-vg-manager"
+
+		// controller reference
+		err := ctrl.SetControllerReference(lvmCluster, ds, r.Scheme)
+		if err != nil {
+			return err
+		}
+		// tolerations
+		ds.Spec.Template.Spec.Tolerations = tolerations
+
+		// nodeSelector if non-nil
+		if nodeSelector != nil {
+			ds.Spec.Template.Spec.Affinity = &corev1.Affinity{
+				NodeAffinity: &corev1.NodeAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: nodeSelector,
+				},
+			}
+		} else {
+			ds.Spec.Template.Spec.Affinity = nil
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		unitLogger.Info("daemonset modified", "operation", result, "name", ds.Name)
+	} else {
+		unitLogger.Info("daemonset unchanged")
+	}
+	return err
+}
+
+func (v vgManager) ensureDeleted(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	return nil
```

Comment:
Add a comment here explaining why nothing needs to be done.

Reply:
done.